Add section on public API deprecation for major NServiceBus upgrades (#7226)
Particular/docs.particular.net: nservicebus/upgrades/index.md

 * [NuGet Package Manager Dialog - Updating a Package](https://docs.microsoft.com/en-us/nuget/tools/package-manager-ui#updating-a-package)
 * [NuGet Package Manager Console - Updating a Package](https://docs.microsoft.com/en-us/nuget/tools/package-manager-console#updating-a-package)

## Public API Deprecation

When preparing for a major NServiceBus upgrade, it is crucial to proactively address any potential breaking changes by thoroughly checking your system for warnings and errors at compile time.

> As part of the [release policy](/nservicebus/upgrades/release-policy.md#deprecation), APIs are marked as obsolete in a major version before removing them in the next. These obsolete APIs will generate build warnings, providing you with a clear indication of the code that needs to be updated. Treating these warnings as errors and resolving them ensures that your system will not fail at runtime after the upgrade.

Reviewer: ```suggestion
As part of the [release policy](/nservicebus/upgrades/release-policy.md#deprecation), APIs are marked as obsolete, with a warning in a minor release and an error in a major release, before removing them in the next major release. These obsolete APIs will generate build warnings or errors, clearly indicating the code that needs to be updated. Treating warnings as errors and resolving them ensures your system will not fail at runtime after the upgrade.
```

Reviewer 2: FYI we are trying to follow this practice with the minor but this practice is not always possible or feasible to follow. I think it would be better to word this as best effort.

> To facilitate a smooth transition, always follow the detailed step-by-step upgrade guide provided for the specific version you are moving to. This guide will not only outline the necessary changes for deprecated APIs but also provide instructions on how to adapt to the new APIs and features, ensuring a successful and stable upgrade.

Author: This is the only guidance I can find addressing [Upgrade Guides](https://github.com/Particular/Strategy/blob/bc47204c2504e7e3f90a7fefe5c05c4e68cce216/definitions/upgrade-guide.md). Do we have anything else? Perhaps this needs to be expanded upon to clarify what exactly is expected in a Upgrade Guide?
